internal/service/alarms/internal/dictionary_definition/dictionary_definition.go

```diff
@@ -0,0 +1,460 @@
+
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, clients.ListRequestTimeout)
+	defer cancel()
+
+	var managedClusters clusterv1.ManagedClusterList
+	err := r.Client.List(ctxWithTimeout, &managedClusters, &crclient.ListOptions{
+		LabelSelector: selector.Add(*versionSelector).Add(*localClusterRequirement),
+		Limit:         1,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error listing managed clusters: %w", err)
+	}
+
+	if len(managedClusters.Items) == 0 {
+		return nil, fmt.Errorf("no managed cluster found with version %s", version)
+	}
+
+	return &managedClusters.Items[0], nil
+}
+
+// getRules gets rules defined within a PrometheusRule resource
+func (r *AlarmDictionaryDefinition) getRules(ctx context.Context, cl crclient.Client) ([]monitoringv1.Rule, error) {
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, clients.ListRequestTimeout)
+	defer cancel()
+
+	var promRules monitoringv1.PrometheusRuleList
+	err := cl.List(ctxWithTimeout, &promRules)
+	if err != nil {
+		return nil, fmt.Errorf("error listing prometheus rules: %w", err)
+	}
+
+	// Extract rules from PrometheusRule list
+	var rules []monitoringv1.Rule
+	for _, promRule := range promRules.Items {
+		for _, group := range promRule.Spec.Groups {
+			for _, rule := range group.Rules {
+				// Only alerting rules are of interest (not recording rules)
+				if rule.Alert != "" {
+					rules = append(rules, rule)
+				}
+			}
+		}
+	}
+
+	return rules, nil
+}
+
+// syncDictionaries synchronizes the alarm dictionaries in the database
+func (r *AlarmDictionaryDefinition) syncDictionaries(ctx context.Context, nodeClusterTypes []clusterserver.NodeClusterType) error {
+	slog.Info("Synchronizing alarm dictionaries in the database")
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Minute) // let's try to finish init with a set time.
+	defer cancel()
+
+	// First clean up outdated dictionaries
+	if err := r.deleteOutdatedDictionaries(ctx, nodeClusterTypes); err != nil {
+		return fmt.Errorf("failed to delete outdated dictionaries: %w", err)
+	}
+
+	// Then process all NodeClusterTypes concurrently
+	if err := r.processNodeClusterTypes(ctx, nodeClusterTypes); err != nil {
+		return fmt.Errorf("failed to process resource types: %w", err)
+	}
+
+	slog.Info("Alarm dictionaries and corresponding AlarmDefinitions synchronized")
+	return nil
+}
+
+// deleteOutdatedDictionaries remove any dictionary that may not be available anymore
+func (r *AlarmDictionaryDefinition) deleteOutdatedDictionaries(ctx context.Context, nodeClusterTypes []clusterserver.NodeClusterType) error {
+	// Delete Dictionaries that do not have a corresponding resource type
+	ids := make([]any, 0, len(nodeClusterTypes))
+	for _, nodeClusterType := range nodeClusterTypes {
+		ids = append(ids, nodeClusterType.NodeClusterTypeId)
+	}
+
+	if err := r.AlarmsRepository.DeleteAlarmDictionariesNotIn(ctx, ids); err != nil {
+		return fmt.Errorf("failed to delete alarm dictionaries: %w", err)
```

Comment:
even if for some reason we cannot delete outdated dictionaries, we can still sync the ones present - We had a warning here before

Reply:
only warning now